dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -308,20 +400,32 @@ ExchangeReceiverBase<RPCContext>::ExchangeReceiverBase(
     size_t source_num_,
     size_t max_streams_,
     const String & req_id,
-    const String & executor_id)
+    const String & executor_id,
+    bool enable_fine_grained_shuffle_)
     : rpc_context(std::move(rpc_context_))
     , source_num(source_num_)
     , max_streams(max_streams_)
     , max_buffer_size(std::max<size_t>(batch_packet_count, std::max(source_num, max_streams_) * 2))
     , thread_manager(newThreadManager())
-    , msg_channel(max_buffer_size)
     , live_connections(source_num)
     , state(ExchangeReceiverState::NORMAL)
     , exc_log(Logger::get("ExchangeReceiver", req_id, executor_id))
     , collected(false)
+    , enable_fine_grained_shuffle(enable_fine_grained_shuffle_)
 {
     try
     {
+        if (enable_fine_grained_shuffle_)
+        {
+            for (size_t i = 0; i < max_streams_; ++i)
+            {
+                msg_channels.push_back(std::make_shared<MPMCQueue<std::shared_ptr<ReceivedMessage>>>(max_buffer_size));
+            }
+        }
+        else
+        {
+            msg_channels.push_back(std::make_shared<MPMCQueue<std::shared_ptr<ReceivedMessage>>>(max_buffer_size));
```

Comment:
Because `MPMCQueue`'s copy constructor is deleted. So we cannot push back `MPMCQueue` directly. So I use `shared_ptr` instead.

Reply:
use `std::vector::emplace_back` instead.

Reply:
Looks like we have to implement copy/move constructor for `MPMCQueue` to use `emplace_back`. I can change to use `unique_ptr`.

Reply:
`MPMCQeue` has no copy/move constructor, so `std::vector::emplace_back()` is cannot compile.
Adding copy/move ctor for `MPMCQeue` looks not suitable, because it's should be thread safe.

So maybe I should `unique_ptr` for `MPMCQueue`.